File: bytedance/Array/033. Search in Rotated Sorted Array.py
# Regular way
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if not nums:
            return -1
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                return mid
            if nums[0] <= nums[mid]:
                if nums[0] <= target < nums[mid]:
                    r = mid - 1
                else:
                    l = mid + 1
            else:
                if nums[mid] < target <= nums[len(nums) - 1]:
                    l = mid + 1
                else:
                    r = mid - 1
        return -1


# Noy easy to understand.
# https://leetcode-cn.com/problems/search-in-rotated-sorted-array/solution/ji-jian-solution-by-lukelee/
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1
        while lo < hi:
            mid = (lo + hi) // 2
            if ((nums[0] > target) ^ (nums[0] > nums[mid]) ^ (target > nums[mid])):
                lo = mid + 1
            else:
                hi = mid
        return lo == hi and lo if nums[lo] == target else -1

File: bytedance/Array/033. Search in Rotated Sorted Array.go
func search(nums []int, target int) int {
    l, r := 0, len(nums) - 1
    for l <= r {
        mid := (l + r) / 2
        if nums[mid] == target {
            return mid
        }
        if nums[mid] >= nums[0] {
            if nums[0] <= target && target < nums[mid] {
                r = mid - 1
            } else {
                l = mid + 1
            }
        } else {
            if nums[mid] < target && target <= nums[len(nums) - 1] {
                l = mid + 1
            } else {
                r = mid - 1
            }
        }
    }

    return -1
}